### ISSUE TITLE
[Bug] Import-DbaCsv 's automap functionality doesn't work with pipe delimited files going into existing tables

### ISSUE THREAD

**Issue by MarkIannucci** [sqlcollaborative/dbatools#6763]
### Environmental information

PowerShell Version        : 5.1.19041.1
dbatools latest installed : 1.0.115
Culture of OS             : en-US

#### SQL Server:

Microsoft SQL Server 2019 (RTM) - 15.0.2000.5 (X64) 	Sep 24 2019 13:48:23 	Copyright (C) 2019 Microsoft Corporation	Developer Edition (64-bit) on Windows 10 Pro for Workstations 10.0 <X64> (Build 19042: ) (Hypervisor)

us_english

### Report

When we tried to use `Import-DbaCsv` to import a pipe delimited file into an existing table, we noticed the automap functionality stopped working and we received this warning:

`Failure | Input string was not in a correct format.`

In the steps to reproduce section, I provided a table definition and example file which reliably reproduce this behavior.

#### Host used

- [X] powershell.exe
- [ ] ISE
- [ ] VS Code
- [ ] Other (please specify)

#### Errors Received

```
writeErrorStream      : True
PSMessageDetails      :
Exception             : System.Exception: Input string was not in a correct format. --->
                        System.Management.Automation.MethodInvocationException: Exception calling "WriteToServer" with
                        "1" argument(s): "The given value of type String from the data source cannot be converted to
                        type bigint of the specified target column." ---> System.InvalidOperationException: The given
                        value of type String from the data source cannot be converted to type bigint of the specified
                        target column. ---> System.FormatException: Failed to convert parameter value from a String to
                        a Int64. ---> System.FormatException: Input string was not in a correct format.
                           at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number,
                        NumberFormatInfo info, Boolean parseDecimal)
                           at System.Number.ParseInt64(String value, NumberStyles options, NumberFormatInfo numfmt)
                           at System.String.System.IConvertible.ToInt64(IFormatProvider provider)
                           at System.Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
                           at System.Data.SqlClient.SqlParameter.CoerceValue(Object value, MetaType destinationType,
                        Boolean& coercedToDataFeed, Boolean& typeChanged, Boolean allowStreaming)
                           --- End of inner exception stack trace ---
                           at System.Data.SqlClient.SqlParameter.CoerceValue(Object value, MetaType destinationType,
                        Boolean& coercedToDataFeed, Boolean& typeChanged, Boolean allowStreaming)
                           at System.Data.SqlClient.SqlBulkCopy.ConvertValue(Object value, _SqlMetaData metadata,
                        Boolean isNull, Boolean& isSqlType, Boolean& coercedToDataFeed)
                           --- End of inner exception stack trace ---
                           at System.Data.SqlClient.SqlBulkCopy.ConvertValue(Object value, _SqlMetaData metadata,
                        Boolean isNull, Boolean& isSqlType, Boolean& coercedToDataFeed)
                           at System.Data.SqlClient.SqlBulkCopy.ReadWriteColumnValueAsync(Int32 col)
                           at System.Data.SqlClient.SqlBulkCopy.CopyColumnsAsync(Int32 col, TaskCompletionSource`1
                        source)
                           at System.Data.SqlClient.SqlBulkCopy.CopyRowsAsync(Int32 rowsSoFar, Int32 totalRows,
                        CancellationToken cts, TaskCompletionSource`1 source)
                           at System.Data.SqlClient.SqlBulkCopy.CopyBatchesAsyncContinued(BulkCopySimpleResultSet
                        internalResults, String updateBulkCommandText, CancellationToken cts, TaskCompletionSource`1
                        source)
                           at System.Data.SqlClient.SqlBulkCopy.CopyBatchesAsync(BulkCopySimpleResultSet
                        internalResults, String updateBulkCommandText, CancellationToken cts, TaskCompletionSource`1
                        source)
                           at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalRestContinuedAsync(BulkCopySimpleR
                        esultSet internalResults, CancellationToken cts, TaskCompletionSource`1 source)
                           at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalRestAsync(CancellationToken cts,
                        TaskCompletionSource`1 source)
                           at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalAsync(CancellationToken ctoken)
                           at System.Data.SqlClient.SqlBulkCopy.WriteRowSourceToServerAsync(Int32 columnCount,
                        CancellationToken ctoken)
                           at System.Data.SqlClient.SqlBulkCopy.WriteToServer(IDataReader reader)
                           at CallSite.Target(Closure , CallSite , Object , Object )
                           --- End of inner exception stack trace ---
                           at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext
                        funcContext, Exception exception)
                           at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame
                        frame)
                           at
                        System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame
                        frame)
                           at
                        System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame
                        frame)
                           --- End of inner exception stack trace ---
TargetObject          :
CategoryInfo          : NotSpecified: (:) [Write-Error], Exception
FullyQualifiedErrorId : dbatools_Import-DbaCsv,Stop-Function
ErrorDetails          : Input string was not in a correct format.
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Stop-Function, C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.115\allcommands.ps1:
                        line 83636
                        at Import-DbaCsv<Process>, C:\Program
                        Files\WindowsPowerShell\Modules\dbatools\1.0.115\allcommands.ps1: line 38794
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {0, 1}
```

#### Steps to Reproduce

[TableDef.TXT](https://github.com/sqlcollaborative/dbatools/files/5133176/TableDef.TXT)
[ReproPipe.TXT](https://github.com/sqlcollaborative/dbatools/files/5133177/ReproPipe.TXT)

Run TableDef.TXT against your tempdb and then execute the `Import-DbaCsv` PowerShell cmdlet:

    Import-DbaCsv -Database "tempdb" -SqlInstance "localhost" -Delimiter "|" -Path .\ReproPipe.TXT -Table "TableDef" -verbose

![image](https://user-images.githubusercontent.com/16762974/91365112-9bfb6300-e7bd-11ea-8633-63c27938f389.png)

#### Expected Behavior

1 row gets inserted

#### Actual Behavior

User receives error.

**Comment by wsmelton** [comment]
I'm not seeing any issue at all with this delimiter on auto-creating the table or importing again with the table already created.

![image](https://user-images.githubusercontent.com/11204251/91305673-fe7a4200-e770-11ea-85b9-ce572bd9029c.png)

Please provide a sample file that has this issue and we can review it more closely.

**Comment by wsmelton** [comment]
Tested in Windows PowerShell and no issue found. There is no reason that I see to require escaping the pipe character for SqlBulkCopy class or the Lumen framework in use.

Please provide much more detail, data requested with our template should be included or future issues may just be closed.

**Comment by MarkIannucci** [comment]
Thank you for taking a look.  I was a bit concerned that my report was going to be too terse.  I wanted to get the PR underway while the code was still in my head.  I can't provide the exact files that produce this on my end due to patient data.  I will clean some of that stuff up tonight and will submit a better repro.

**Comment by MarkIannucci** [comment]
I updated my first comment with the additional information @wsmelton requested.  

I'm not entirely sure why the code as written intermittently works.  I do know that the solution I propose in #6764 would only be intermittently executed because the [automap function](https://github.com/sqlcollaborative/dbatools/blob/f8f4897a0fe4e644416becaceb2140002499cb30/functions/Import-DbaCsv.ps1#L561) gets skipped [if the user uses](https://github.com/sqlcollaborative/dbatools/blob/f8f4897a0fe4e644416becaceb2140002499cb30/functions/Import-DbaCsv.ps1#L556) the `-AutoCreateTable` switch.  I suspect the same holds if the table and csv have the exact same columns names and order.  

Additionally, I noticed that if I write the following in powershell, I get an incorrect split:

    $myline = 'foo|bar|boo'
    $myline -split '|'

However if I escape the pipe, I get the split we expect:

![image](https://user-images.githubusercontent.com/16762974/91366180-483e4900-e7c0-11ea-8a39-9a67030c4f50.png)

**Comment by MarkIannucci** [comment]
In writing my post above, I realized my repro steps could be further simplified.  I updated the example a few moments ago.

**Comment by wsmelton** [comment]
Use of `-split` cannot be compared to CSV data parsing, they are not the same processing from .NET perspective. Split splits the string as a whole and is not parsing for data that CSV structure contains.

You can make any character you want and mix and match and split will still spit out each individual character:
![image](https://user-images.githubusercontent.com/11204251/91372957-bc85e600-e7da-11ea-9fc3-7883da34c275.png)

**Comment by wsmelton** [comment]
This could be fixed by cleaning up some of the code in how it is trying to detect the delimiter and not make it work as a string 
So this section of code:
https://github.com/sqlcollaborative/dbatools/blob/f8f4897a0fe4e644416becaceb2140002499cb30/functions/Import-DbaCsv.ps1#L393-L404

Needs to be adjusted to (1) better detect if the pipeling delimiter is correct without having to import a complete file (large files would cause perf issues). Then (2) we can re-work this to include getting the columns of the file so that address getting this section to not use Split method:

https://github.com/sqlcollaborative/dbatools/blob/f8f4897a0fe4e644416becaceb2140002499cb30/functions/Import-DbaCsv.ps1#L556-L572

**Comment by MarkIannucci** [comment]
I'm sorry, I don't know how to fix the potential performance problem, and I cannot devote time to complete the research necessary to do that.  We are not having any trouble processing a 2 GB file with more than a million rows.  I have to leave that work to someone else.

I understand your concern about my PR. `import-dbacsv` has a small address space of characters that are limited to the following:
https://github.com/sqlcollaborative/dbatools/blob/f8f4897a0fe4e644416becaceb2140002499cb30/functions/Import-DbaCsv.ps1#L26

`-split` has a `"SimpleMatch"` option [parameter](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_split?view=powershell-7#parameters) which we can use to get it to stop using regular expressions.  

![image](https://user-images.githubusercontent.com/16762974/91446466-bfadc000-e834-11ea-8c1a-329360cbbb15.png)

We can use this to fix the bug which I've done in PR #6769

**Comment by wsmelton** [comment]
The use of split needs to be taken out of the process as a whole. There should be better methods for handling this that don't depend upon string manipulation.

Issues are meant for discussion to reach the final solution before submitting PRs. The original poster has the option to implement that solution once it has been agreed/approved but is not required. Other contributors can come in and see the discussion and approval and do the work as well.
